golang/cosmos/x/vstorage/keeper/grpc_query.go

```diff
@@ -30,6 +38,209 @@ func (k Querier) Data(c context.Context, req *types.QueryDataRequest) (*types.Qu
+
+const (
+	// Media types.
+	JSONLines = "JSON Lines"
+
+	// CapData transformation formats.
+	FormatCapDataFlat = "flat"
+
+	// CapData remotable value formats.
+	FormatRemotableAsObject = "object"
+	FormatRemotableAsString = "string"
+)
+
+var capDataResponseMediaTypes = map[string]string{
+	JSONLines: JSONLines,
+	// Default to JSON Lines.
+	"": JSONLines,
+}
+var capDataTransformationFormats = map[string]string{
+	FormatCapDataFlat: FormatCapDataFlat,
+	// Default to no transformation.
+	"": "",
+}
+var capDataRemotableValueFormats = map[string]string{
+	FormatRemotableAsObject: FormatRemotableAsObject,
+	FormatRemotableAsString: FormatRemotableAsString,
+	// No default.
+}
+
+// flatten converts data into a flat structure in which each deep leaf entry is replaced with
+// a top-level entry having the same value but a key generated by joining the keys on its path
+// with separating dashes.
+// For example,
+// ```
+// { "contacts": [
+//
+//	{ "name": "Alice", "email": "a@example.com" },
+//	{ "name": "Bob", "email": "b@example.com" }
+//
+// ] }
+// ```
+// becomes
+// ```
+//
+//	{
+//	  "contacts-0-name": "Alice",
+//	  "contacts-0-email": "a@example.com",
+//	  "contacts-1-name": "Bob",
+//	  "contacts-1-email": "b@example.com"
+//	}
+//
+// ```
+// cf. https://github.com/Agoric/agoric-sdk/blob/6e5b422b80e47c4dac151404f43faea5ab41e9b0/scripts/get-flattened-publication.sh
+func flatten(input interface{}, output map[string]interface{}, key string, top bool) error {
+	// Act on the raw representation of a Remotable.
+	if capdata, ok := input.(*capdata.CapdataRemotable); ok {
+		repr, _ := json.Marshal(capdata)
+		var replacement interface{}
+		if err := json.Unmarshal(repr, &replacement); err != nil {
+			return err
+		}
+		input = replacement
+	}
+
+	childKeyPrefix := key
+	if !top {
+		childKeyPrefix = childKeyPrefix + "-"
+	}
+	if arr, ok := input.([]interface{}); ok {
+		for i, v := range arr {
+			if err := flatten(v, output, childKeyPrefix+fmt.Sprintf("%d", i), false); err != nil {
+				return err
+			}
+		}
+	} else if obj, ok := input.(map[string]interface{}); ok {
+		for k, v := range obj {
+			if err := flatten(v, output, childKeyPrefix+k, false); err != nil {
+				return err
+			}
+		}
+	} else {
+		if _, has := output[key]; has {
+			return fmt.Errorf("key conflict: %q", key)
+		}
+		output[key] = input
+	}
+	return nil
+}
+
+// capdataBigintToDigits represents a bigint as a string consisting of
+// an optional "-" followed by a sequence of digits with no extraneous zeroes
+// (e.g., "0" or "-40").
+func capdataBigintToDigits(bigint *capdata.CapdataBigint) interface{} {
+	return bigint.Normalized
+}
+
+// capdataRemotableToString represents a Remotable as a bracketed string
+// containing its alleged name (e.g., "[Foo {}]").
+func capdataRemotableToString(r *capdata.CapdataRemotable) interface{} {
+	iface := "Remotable"
+	if r.Iface != nil || *r.Iface != "" {
+		iface = *r.Iface
+	}
+	return fmt.Sprintf("[%s {}]", iface)
+}
```

Comment:
In order to prevent clients from relying on the `Iface` (aka debug name) for correctness, the rendering of a remotable _must_ preserve the Id. Including the `Iface` at all is risky.

Reply:
Per #7581, this attempts to align with `agoric follow -o jsonlines`:
> ```sh
> agoric --sdk follow :published.vaultFactory.manager0.metrics -B https://devnet.agoric.net/network-config --proof none -o jsonlines | jq .
> {
>   "liquidatingCollateral": {
>     "brand": "[Alleged: SEVERED: IbcATOM brand {}]",
>     "value": "0"
>   },
>   …
> ```

Who is the right group to discuss well-motivated deviation? @arirubinstein?